user_sync/connector/directory_ldap.py

```diff
@@ -130,51 +139,60 @@ def load_users_and_groups(self, groups, extended_attributes, all_users):
         :rtype (bool, iterable(dict))
         """
         options = self.options
-        all_users_filter = six.text_type(options['all_users_filter'])
-        group_member_filter_format = six.text_type(options['group_member_filter_format'])
+        grouped_user_records = {}
+        if self.two_steps_lookup:
+            group_member_attribute_name = six.text_type(options['group_member_attribute_name'])
+
+        # save all the users to memory for faster 2-steps lookup or all_users process
+        if all_users or self.two_steps_lookup:
+            all_users_filter = six.text_type(options['all_users_filter'])
+            try:
+                all_users_records = dict(self.iter_users(all_users_filter, extended_attributes))
```

Comment:
Load all users for comparison fails (due to max LDAP query limit for regular users) or timeout (for Directory Manager) for large Directory such as 500K+ users.  Please include option to do per group based users lookup instead of loading full set.